provider/oracle/storage_volumes.go

```diff
@@ -0,0 +1,534 @@
+	details, err = s.api.CreateStorageVolume(params)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Infof("waiting for resource %v", details.Name)
+
+	// wait for the newly created volume to reach "Online" status
+	if err := s.waitForResourceStatus(
+		s.fetchVolumeStatus,
+		string(details.Name),
+		string(ociCommon.VolumeOnline), 5*time.Minute); err != nil {
+		return nil, errors.Trace(err)
+	}
+	volume := &storage.Volume{
+		p.Tag,
+		storage.VolumeInfo{
+			VolumeId: details.Name,
+			// the API returns the size of the volume in bytes.
+			// convert to GiB
+			Size:       uint64(details.Size) / 1024 / 1024 / 1024,
+			Persistent: true,
+		},
+	}
+	logger.Infof("returning volume details: %v", volume)
+	return volume, nil
+}
+
+// CreateVolumes is specified on the storage.VolumeSource interface
+func (s *oracleVolumeSource) CreateVolumes(params []storage.VolumeParams) ([]storage.CreateVolumesResult, error) {
+	if params == nil {
+		return []storage.CreateVolumesResult{}, nil
+	}
+	results := make([]storage.CreateVolumesResult, len(params))
+	for i, volume := range params {
+		vol, err := s.createVolume(volume)
+		if err != nil {
+			results[i].Error = errors.Trace(err)
+			continue
+		}
+		results[i].Volume = vol
+	}
+	return results, nil
+}
+
+// fetchVolumeStatus polls the status of a volume and returns true if the current status
+// coincides with the desired status
+func (s *oracleVolumeSource) fetchVolumeStatus(name, desiredStatus string) (complete bool, err error) {
+	details, err := s.api.StorageVolumeDetails(name)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	if details.Status == ociCommon.VolumeError {
+		return false, errors.Errorf("volume entered error state: %q", details.Status_detail)
+	}
+	return string(details.Status) == desiredStatus, nil
+}
+
+// fetchVolumeAttachmentStatus polls the status of a volume attachment and returns true if the current status
+// coincides with the desired status
+func (s *oracleVolumeSource) fetchVolumeAttachmentStatus(name, desiredStatus string) (bool, error) {
+	details, err := s.api.StorageAttachmentDetails(name)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	return string(details.State) == desiredStatus, nil
+}
+
+// waitForResourceStatus will ping the resource until the fetch function returns true,
+// the timeout is reached, or an error occurs.
+func (o *oracleVolumeSource) waitForResourceStatus(
+	fetch func(name string, desiredStatus string) (complete bool, err error),
+	name, state string, timeout time.Duration) error {
+	errChan := make(chan error)
+	done := make(chan bool)
+
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			default:
+				done, err := fetch(name, state)
+				if err != nil {
+					errChan <- err
+					return
+				}
+				if done {
+					errChan <- nil
+					return
+				}
+				time.Sleep(2 * time.Second)
+			}
+		}
+	}()
+
+	select {
+	case err := <-errChan:
+		return err
+	case <-time.After(timeout):
```

Comment:
I see no benefit in using this here. It looks like a big requirement for a simple task. Somewhat like killing a mosquito with a hammer. But if it's a project requirement....sure. Will change.